lib/datadog/tracing/trace_segment.rb

```diff
@@ -54,18 +54,18 @@ def initialize(
 
         # Set well-known tags
         self.agent_sample_rate = agent_sample_rate
-        self.hostname = (hostname && hostname.dup)
-        self.lang = (lang && lang.dup)
-        self.name = (name && name.dup)
-        self.origin = (origin && origin.dup)
+        self.hostname = (hostname.frozen? ? hostname : hostname.dup)
+        self.lang = (lang.frozen? ? lang : lang.dup)
+        self.name = (name.frozen? ? name : name.dup)
+        self.origin = (origin.frozen? ? origin : origin.dup)
         self.process_id = process_id
         self.rate_limiter_rate = rate_limiter_rate
-        self.resource = (resource && resource.dup)
+        self.resource = (resource.frozen? ? resource : resource.dup)
         self.rule_sample_rate = rule_sample_rate
-        self.runtime_id = (runtime_id && runtime_id.dup)
+        self.runtime_id = (runtime_id.frozen? ? runtime_id : runtime_id.dup)
```

Comment:
Since we're always the ones that should provide the `runtime_id`, have you considered:

a) Changing the `Datadog::Core::Environment::Identity.id` helper to always freeze its result
b) Just calling `runtime_id.freeze` directly?

In particular for `b)` it seems reasonable to assume that this value will come from us, and if a very very very very very very very very advanced user is overriding it then it doesn't seem that bad to require them to live with the fact that we call `freeze` here.

(`b)` also applies to `lang` I think...)

Reply:
I'm freezing the result of `Datadog::Core::Environment::Identity.id`